File: PiApproximationByThrowingDarts.hs
module PiApproximationByThrowingDarts where

import System.Environment (getArgs)
import System.Random (randomRIO)

piApprox :: Double -> IO Double
piApprox sampling = do
  count <- mapM dart [1..sampling]
  return $ sum count / sampling * 4
  where
    dart _ = do
      x <- randomRIO (0, 1) :: IO Double
      y <- randomRIO (0, 1) :: IO Double
      return $ if x ^ 2 + y ^ 2 < 1 then 1 else 0

main :: IO ()
main = do
  args <- getArgs
  let sampling = if length args == 1
                   then read $ head args :: Double
                   else 1000000
  result <- piApprox sampling
  putStrLn $ "π ≒ " ++ show result

File: pi_approximation_by_throwing_darts.rb
module PiApproximationByThrowingDarts
  module_function

  def pi_approx(sampling)
    dart = lambda do |n|
      x, y = Random.rand(1.0), Random.rand(1.0)
      x ** 2 + y ** 2 < 1 ? 1 : 0
    end
    count = (1..sampling).collect(&dart).inject(:+)
    count.to_f / sampling * 4
  end
end

if __FILE__ == $0
  sampling = ARGV.length == 1 ? ARGV[0].to_i : 1000000
  puts "π ≒ #{PiApproximationByThrowingDarts.pi_approx(sampling)}"
end
